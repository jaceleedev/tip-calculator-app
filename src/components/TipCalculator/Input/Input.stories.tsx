import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

import Input from './Input';

const meta = {
  component: Input,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    iconConfig: {
      description:
        'Icon configuration including the source, dimensions, and margin settings',
      control: {
        type: 'object',
      },
    },
    isValid: {
      description: 'Indicates whether the input value is valid',
      control: 'boolean',
    },
    errorMessage: {
      description: 'Message displayed when the input is invalid',
      control: 'text',
    },
    value: {
      description: 'The current value of the input field.',
      control: 'text',
    },
    placeholder: {
      description: 'Placeholder text displayed when the input is empty',
      control: 'text',
    },
    onChange: {
      description: 'Event handler for input value changes',
    },
  },
} satisfies Meta<typeof Input>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Dollar: Story = {
  decorators: [
    (Story) => (
      <div style={{ width: '311px' }}>
        <Story />
      </div>
    ),
  ],
  args: {
    iconConfig: {
      src: '/images/icon-dollar.svg',
      width: 11,
      height: 17,
      marginLeft: 19.26,
    },
    placeholder: '0',
    isValid: true,
  },
};

export const People: Story = {
  decorators: [
    (Story) => (
      <div style={{ width: '311px' }}>
        <Story />
      </div>
    ),
  ],
  args: {
    iconConfig: {
      src: '/images/icon-person.svg',
      width: 13,
      height: 16,
      marginLeft: 17,
    },
    placeholder: '0',
    isValid: true,
  },
};

export const Custom: Story = {
  decorators: [
    (Story) => (
      <div style={{ width: '147.5px' }}>
        <Story />
      </div>
    ),
  ],
  args: {
    placeholder: 'Custom',
  },
};

export const Valid: Story = {
  decorators: [
    (Story) => (
      <div style={{ width: '311px' }}>
        <Story />
      </div>
    ),
  ],
  args: {
    iconConfig: {
      src: '/images/icon-person.svg',
      width: 13,
      height: 16,
      marginLeft: 17,
    },
    isValid: true,
    value: 5,
    onChange: fn(),
  },
  parameters: {
    pseudo: { focus: true },
  },
};

export const Invalid: Story = {
  decorators: [
    (Story) => (
      <div style={{ width: '311px' }}>
        <Story />
      </div>
    ),
  ],
  args: {
    iconConfig: {
      src: '/images/icon-person.svg',
      width: 13,
      height: 16,
      marginLeft: 17,
    },
    isValid: false,
    errorMessage: `Can't be zero`,
    value: 0,
    onChange: fn(),
  },
  parameters: {
    pseudo: { focus: true },
  },
};
