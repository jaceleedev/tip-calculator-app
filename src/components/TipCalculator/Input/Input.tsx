import { clsx } from 'clsx';
import Image from 'next/image';
import { InputHTMLAttributes } from 'react';

interface IconConfig {
  src: string;
  width: number;
  height: number;
  marginLeft: number;
}

interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  iconConfig?: IconConfig;
  isValid?: boolean;
  errorMessage?: string;
}

function Input({
  iconConfig,
  isValid,
  errorMessage,
  ...props
}: Readonly<InputProps>) {
  const inputClassName = clsx(
    'flex items-center text-primary text-very-dark-cyan text-right w-full h-12 pr-[17px] rounded-[5px] border-2 border-solid border-transparent outline-none bg-very-light-grayish-cyan',
    {
      'placeholder:text-opacity-35': iconConfig,
      'placeholder:text-medium-dark-grayish-cyan': !iconConfig,
      'focus:border-strong-cyan caret-strong-cyan': isValid,
      'focus:border-strong-orange-red caret-strong-orange-red': !isValid,
    }
  );

  return (
    <div className="relative">
      {iconConfig && <InputIcon iconConfig={iconConfig} />}
      <input type="text" className={inputClassName} {...props} />
      {!isValid && errorMessage && <ErrorMessage message={errorMessage} />}
    </div>
  );
}

function InputIcon({ iconConfig }: Readonly<{ iconConfig: IconConfig }>) {
  return (
    <div
      style={{ marginLeft: `${iconConfig.marginLeft}px` }}
      className={`absolute inset-y-0 left-0 flex items-center`}
    >
      <Image
        src={iconConfig.src}
        alt=""
        width={iconConfig.width}
        height={iconConfig.height}
        priority
      />
    </div>
  );
}

function ErrorMessage({ message }: Readonly<{ message: string }>) {
  return (
    <span className="absolute right-0 -top-[30px] text-error text-strong-orange-red">
      {message}
    </span>
  );
}

export default Input;
