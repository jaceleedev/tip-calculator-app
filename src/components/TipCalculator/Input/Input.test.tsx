import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import Input from './Input';

expect.extend(toHaveNoViolations);

describe('Input', () => {
  // 1. Rendering test
  it('renders correctly', () => {
    render(
      <Input
        id="rendering-test"
        placeholder="rendering test"
        isValid={true}
        errorMessage="rendering message"
      />
    );
    const input = screen.getByPlaceholderText('rendering test');
    expect(input).toBeDefined();
  });

  // 2. Props test
  it('applies correct attributes', () => {
    render(
      <Input
        id="props-test"
        placeholder="props test"
        type="text"
        isValid={true}
        errorMessage="props test"
      />
    );
    const input = screen.getByPlaceholderText('props test');
    expect(input).toHaveProperty('id', 'props-test');
    expect(input).toHaveProperty('type', 'text');
    expect(input).toHaveProperty('placeholder', 'props test');
  });

  // 3. Styles test
  it('applies the correct styles', () => {
    render(
      <Input
        id="styles test"
        placeholder="styles test"
        isValid={true}
        errorMessage="styles message"
      />
    );
    const input = screen.getByPlaceholderText('styles test');
    expect(input.className).toContain('text-primary');
    expect(input.className).toContain('text-very-dark-cyan');
    expect(input.className).toContain('w-full');
    expect(input.className).toContain('h-12');
  });

  // 4. Error styles test
  it('applies invalid styles when isValid prop is false', () => {
    render(
      <Input
        id="error styles test"
        placeholder="error styles test"
        isValid={false}
        errorMessage="error styles message"
      />
    );
    const input = screen.getByPlaceholderText('error styles test');
    expect(input.className).toContain('focus:border-strong-orange-red');
    expect(input.className).toContain('caret-strong-orange-red');
    expect(screen.getByText('error styles message')).toBeDefined();
  });

  // 5. State change test
  it('shows error message when invalid', () => {
    render(
      <Input
        id="invalid test"
        placeholder="invalid test"
        isValid={false}
        errorMessage="invalid message"
      />
    );
    const errorMessage = screen.getByText('invalid message');
    expect(errorMessage).toBeDefined();
  });

  // 6. Behavior test
  it('calls onChange when input value changes', () => {
    const handleChange = vi.fn();
    render(
      <Input
        id="behavior test"
        placeholder="behavior test"
        isValid={true}
        errorMessage="behavior message"
        onChange={handleChange}
      />
    );
    const input = screen.getByPlaceholderText('behavior test');
    fireEvent.change(input, { target: { value: 'test' } });
    expect(handleChange).toHaveBeenCalledOnce();
  });

  // 7. Accessibility test
  it('has no accessibility violations', async () => {
    const { container } = render(
      <Input
        id="accessibility"
        placeholder="accessibility test"
        isValid={true}
        errorMessage="accessibility message"
      />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  // 8. Snapshot test
  it('matches snapshot', () => {
    const { asFragment } = render(
      <Input
        id="snapshot"
        placeholder="snapshot test"
        isValid={true}
        errorMessage="snapshot message"
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
