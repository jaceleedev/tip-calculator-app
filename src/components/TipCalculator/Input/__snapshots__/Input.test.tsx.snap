// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <input
      class="flex items-center text-primary text-very-dark-cyan text-right w-full h-12 pr-[17px] rounded-[5px] border-2 border-solid border-transparent outline-none bg-very-light-grayish-cyan placeholder:text-medium-dark-grayish-cyan focus:border-strong-cyan caret-strong-cyan"
      id="snapshot"
      placeholder="snapshot test"
      type="text"
    />
  </div>
</DocumentFragment>
`;
